incubator/nodejs-express/templates/kafka/app.js

```diff
@@ -31,8 +31,15 @@ module.exports = (options) => {
     // 'log.connection.close' : false,
   }
 
-  Produce(Kafka, config, log, app);
-  Consume(Kafka, brokers, log);
+  producer = Produce(Kafka, config, log, app);
+  consumer = Consume(Kafka, brokers, log);
+
+  // Disconnect our Kafka clients when the HTTP server is stopped.
+  options.server.on('close', function () {
+    log.info('Disconnecting Kafka clients');
+    producer.disconnect();
```

Comment:
```suggestion
    if (producer._isConnected)
      producer.disconnect();
```

just in case if the close call comes too fast before the producer is setup? I am not sure if this is possible in real world cases, but I got `Producer not connected` error while trying to isolate the issue, with a similar change

Reply:
I don't like the look of checking an internal variable called `_isConnected` - this looks like a high risk of breakage in the future?

As I understand it, if the producer / consumer are currently connected, then a call to `disconnect()` will disconnect them.  If they failed to connect for some reason, this call would be unnecessary (but do no harm since we're shutting down anyway).

The situation that might be a problem would be if the clients were still in the process of connecting, and were going to connect successfully, but we called disconnect before they finished.  I don't know how to test that, and I didn't find any documentation on how you might handle that.  I can't imagine the behaviour would be any worse than the current (indefinitely hanging) behaviour though.

Reply:
I don't like the use of internal variable either. I saw a crash, and I figured the root cause is a race condition, and hence suggested. I am not blocking the PR on this.